storage/src/vespa/storage/persistence/filestorage/service_layer_host_info_reporter.cpp

```diff
@@ -0,0 +1,75 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
+#include "service_layer_host_info_reporter.h"
+#include <vespa/storage/common/nodestateupdater.h>
+#include <cmath>
+
+namespace storage {
+
+using Object = vespalib::JsonStream::Object;
+using End = vespalib::JsonStream::End;
+
+namespace {
+
+void write_usage(vespalib::JsonStream& output, const vespalib::string &label, double value)
+{
+    output << label << Object();
+    output << "usage" << value;
+    output << End();
+}
+
+bool want_immediate_report(const spi::ResourceUsage& old_resource_usage, const spi::ResourceUsage& resource_usage)
+{
+    auto disk_usage_diff = fabs(resource_usage.get_disk_usage() - old_resource_usage.get_disk_usage());
+    auto memory_usage_diff = fabs(resource_usage.get_memory_usage() - old_resource_usage.get_memory_usage());
+    return (disk_usage_diff > 0.01 || memory_usage_diff > 0.01);
+}
+
+}
+
+ServiceLayerHostInfoReporter::ServiceLayerHostInfoReporter(NodeStateUpdater& node_state_updater)
+    : HostReporter(),
+      spi::ResourceUsageListener(),
+      _node_state_updater(node_state_updater),
+      _lock(),
+      _old_resource_usage()
+{
+}
+
+ServiceLayerHostInfoReporter::~ServiceLayerHostInfoReporter()
+{
+    spi::ResourceUsageListener::reset(); // detach
+}
+
+void
+ServiceLayerHostInfoReporter::update_resource_usage(const spi::ResourceUsage& resource_usage)
+{
+    bool immediate_report = want_immediate_report(_old_resource_usage, resource_usage);
+    if (immediate_report) {
+        _old_resource_usage = resource_usage;
```

Comment:
Why are we not required to take the lock here?

Reply:
The lock is to prevent race against calls to report. _old_resource_usage is not used by report.